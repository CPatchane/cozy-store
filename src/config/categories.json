[
  "banking",
  "cozy",
  "energy",
  "health",
  "host_provider",
  "insurance",
  "isp",
  "mes_infos",
  "partners",
  "productivity",
  "ptnb",
  "public_service",
  "shopping",
  "social",
  "telecom",
  "transport"
]
