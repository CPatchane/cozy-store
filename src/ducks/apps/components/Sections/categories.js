/**
 * Deal with app categories.
 *
 * Category objects have the following attributes
 *
 * - value: Slug of the category
 * - label: Translated name of the category
 * - secondary: Whether to be displayed in the select
 */

import { APP_TYPE } from './constants'

/**
 * Like groupBy except that grouper must return an array values.
 * Thus, an object can be in several groups.
 * If the grouper returns a falsy value, an empty list is assumed.
 */
const multiGroupBy = (iter, grouper) => {
  const groups = {}
  for (const obj of iter) {
    const values = grouper(obj) || []
    values.forEach(v => {
      if (!groups.hasOwnProperty(v)) groups[v] = []
      groups[v].push(obj)
    })
  }
  return groups
}

const getAppCategory = app => app.categories
export const groupApps = apps => multiGroupBy(apps, getAppCategory)

// Function to sort categories objects
//
// Alphabetical sort on label except for
//   - 'all' value always at the beginning
//   - 'others' value always at the end
export const sorter = (categoryA, categoryB) => {
  return (
    (categoryA.value === 'all' && -1) ||
    (categoryB.value === 'all' && 1) ||
    (categoryA.value === 'others' && 1) ||
    (categoryB.value === 'others' && -1) ||
    categoryA.label > categoryB.label
  )
}

export const addLabel = (cat, t) => ({
  ...cat,
  label: t(`app_categories.${cat.value}`)
})

// Returns unlabelised/unsorted categories from a list of apps
export const generateOptionsFromApps = (apps, includeAll = false) => {
  if (!apps || !apps.length) return []
  let appsCategories = includeAll
    ? [
        {
          value: 'all',
          secondary: false
        }
      ]
    : []

  for (const type of [APP_TYPE.WEBAPP, APP_TYPE.KONNECTOR]) {
    const catApps = groupApps(apps.filter(a => a.type === type))
    if (type === APP_TYPE.KONNECTOR) {
      // add konnectors at the end
      appsCategories.push({
        value: 'konnectors',
        secondary: false
      })
    }
    const categories = Object.keys(catApps).map(cat => ({
      value: cat,
      type: type,
      secondary: type === APP_TYPE.KONNECTOR
    }))
    appsCategories = appsCategories.concat(categories)
  }

  return appsCategories
}
