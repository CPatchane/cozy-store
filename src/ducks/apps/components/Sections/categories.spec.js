'use strict'

/* eslint-env jest */

import * as catUtils from './categories'

import { tMock } from 'jestLib/I18n'
import mockApps from 'ducks/apps/_mockApps'

describe('groupApps', () => {
  it('should return apps sorted in a dictionnary by categories', () => {
    expect(catUtils.groupApps(mockApps)).toMatchSnapshot()
  })
})

describe('sorter', () => {
  const prepareCategories = categoryNames =>
    categoryNames.map(cat =>
      catUtils.addLabel(
        {
          value: cat
        },
        tMock
      )
    )

  it('should return a categories list alphabetically except for all and others', () => {
    const catList = prepareCategories([
      'cozy',
      'all',
      'isp',
      'others',
      'telecom'
    ])
    expect(catList.sort(catUtils.sorter).map(x => x.value)).toMatchSnapshot()

    // test with all and others at the border
    const catList2 = prepareCategories(['all', 'cozy', 'telecom', 'others'])
    expect(catList2.sort(catUtils.sorter).map(x => x.value)).toMatchSnapshot()
  })
})

describe('generateOptionsFromApps', () => {
  const addLabel = x => catUtils.addLabel(x, tMock)
  it('should return a list of categories options for a select input based on the apps list', () => {
    expect(
      catUtils.generateOptionsFromApps(mockApps, false, addLabel)
    ).toMatchSnapshot()
  })

  it('should return include the all categories if includeAll option true', () => {
    expect(
      catUtils.generateOptionsFromApps(mockApps, true, addLabel)
    ).toMatchSnapshot()
  })

  it('should return an empty list if empty apps list provided', () => {
    expect(
      catUtils.generateOptionsFromApps([], false, addLabel)
    ).toMatchSnapshot()
  })

  it('should return an empty list if no apps provided', () => {
    expect(
      catUtils.generateOptionsFromApps(null, false, addLabel)
    ).toMatchSnapshot()
  })
})
