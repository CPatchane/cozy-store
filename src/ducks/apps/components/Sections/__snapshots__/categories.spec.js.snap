// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateOptionsFromApps should return a list of categories options for a select input based on the apps list 1`] = `
Array [
  Object {
    "label": "Partners",
    "secondary": false,
    "type": "webapp",
    "value": "partners",
  },
  Object {
    "label": "The essentials",
    "secondary": false,
    "type": "webapp",
    "value": "cozy",
  },
  Object {
    "label": "Others",
    "secondary": false,
    "type": "webapp",
    "value": "others",
  },
  Object {
    "label": "Services",
    "secondary": false,
    "value": "konnectors",
  },
  Object {
    "label": "ISP",
    "secondary": true,
    "type": "konnector",
    "value": "isp",
  },
  Object {
    "label": "Telecom",
    "secondary": true,
    "type": "konnector",
    "value": "telecom",
  },
  Object {
    "label": "Transportation",
    "secondary": true,
    "type": "konnector",
    "value": "transport",
  },
]
`;

exports[`generateOptionsFromApps should return an empty list if empty apps list provided 1`] = `Array []`;

exports[`generateOptionsFromApps should return an empty list if no apps provided 1`] = `Array []`;

exports[`generateOptionsFromApps should return include the all categories if includeAll option true 1`] = `
Array [
  Object {
    "secondary": false,
    "value": "all",
  },
  Object {
    "label": "Partners",
    "secondary": false,
    "type": "webapp",
    "value": "partners",
  },
  Object {
    "label": "The essentials",
    "secondary": false,
    "type": "webapp",
    "value": "cozy",
  },
  Object {
    "label": "Others",
    "secondary": false,
    "type": "webapp",
    "value": "others",
  },
  Object {
    "label": "Services",
    "secondary": false,
    "value": "konnectors",
  },
  Object {
    "label": "ISP",
    "secondary": true,
    "type": "konnector",
    "value": "isp",
  },
  Object {
    "label": "Telecom",
    "secondary": true,
    "type": "konnector",
    "value": "telecom",
  },
  Object {
    "label": "Transportation",
    "secondary": true,
    "type": "konnector",
    "value": "transport",
  },
]
`;

exports[`groupApps should return apps sorted in a dictionnary by categories 1`] = `
Object {
  "cozy": Array [
    Object {
      "categories": Array [
        "cozy",
      ],
      "developer": Object {
        "name": "Cozy",
      },
      "editor": "Cozy",
      "icon": "<svg></svg>",
      "installed": true,
      "isInRegistry": true,
      "label": 1,
      "name": "Collect",
      "name_prefix": "Cozy",
      "permissions": Object {
        "mock": Object {
          "type": "io.mock.doctype",
        },
        "mock2": Object {
          "type": "io.mock.doctype2",
        },
      },
      "related": "http://collect.cozy.mock/",
      "slug": "collect",
      "tags": Array [
        "konnector",
        "collect",
        "bills",
        "providers",
        "files",
      ],
      "type": "webapp",
      "uninstallable": false,
      "version": "3.0.0",
      "versions": Object {
        "beta": Array [
          "3.0.0",
        ],
        "dev": Array [
          "3.0.0",
        ],
        "stable": Array [
          "3.0.0",
        ],
      },
    },
    Object {
      "categories": Array [
        "cozy",
      ],
      "developer": Object {
        "name": "Cozy",
      },
      "editor": "Cozy",
      "icon": "<svg></svg>",
      "installed": true,
      "name": "Drive",
      "name_prefix": "Cozy",
      "permissions": Object {
        "mock2": Object {
          "type": "io.mock.doctype2",
        },
      },
      "related": "http://drive.cozy.mock/",
      "slug": "drive",
      "tags": Array [
        "search",
        "files",
        "folders",
      ],
      "type": "webapp",
      "uninstallable": false,
      "version": "3.0.0-beta89bnhj3993",
    },
    Object {
      "categories": Array [
        "cozy",
      ],
      "developer": Object {
        "name": "Cozy",
      },
      "editor": "Cozy",
      "icon": "<svg></svg>",
      "isInRegistry": true,
      "locales": Object {
        "en": Object {
          "long_description": "A long description finally short",
        },
      },
      "name": "Photos",
      "name_prefix": "Cozy",
      "slug": "photos",
      "type": "webapp",
      "uninstallable": true,
      "versions": Object {
        "beta": Array [
          "3.0.0",
        ],
        "dev": Array [
          "3.0.0",
        ],
        "stable": Array [
          "3.0.0",
        ],
      },
    },
  ],
  "isp": Array [
    Object {
      "categories": Array [
        "isp",
        "telecom",
      ],
      "developer": Object {
        "name": "Cozy",
      },
      "icon": "<svg></svg>",
      "isInRegistry": true,
      "name": "Bouilligue",
      "slug": "konnector-bouilligue",
      "type": "konnector",
      "uninstallable": true,
      "versions": Object {
        "beta": Array [
          "0.1.0",
        ],
        "dev": Array [
          "0.1.0",
        ],
        "stable": Array [
          "0.1.0",
        ],
      },
    },
  ],
  "others": Array [
    Object {
      "categories": Array [
        "others",
      ],
      "developer": Object {
        "name": "Cozy",
      },
      "editor": "Cozy",
      "icon": "<svg></svg>",
      "isInRegistry": true,
      "locales": Object {
        "name": "For Dev Only",
      },
      "name": "DevOnly",
      "slug": "devonly",
      "type": "webapp",
      "uninstallable": true,
      "versions": Object {
        "beta": Array [],
        "dev": Array [
          "1.0.0-betaojdkehy989ekhflldh",
        ],
        "stable": Array [],
      },
    },
  ],
  "partners": Array [
    Object {
      "availableVersion": "3.0.0",
      "categories": Array [
        "partners",
      ],
      "icon": "<svg></svg>",
      "installed": true,
      "isInRegistry": true,
      "name": "Tasky",
      "slug": "tasky",
      "type": "webapp",
      "uninstallable": true,
      "versions": Object {
        "beta": Array [
          "3.0.0",
          "1.0.0",
        ],
        "dev": Array [
          "3.0.0",
          "1.0.0",
        ],
        "stable": Array [
          "3.0.0",
          "1.0.0",
        ],
      },
    },
  ],
  "telecom": Array [
    Object {
      "categories": Array [
        "isp",
        "telecom",
      ],
      "developer": Object {
        "name": "Cozy",
      },
      "icon": "<svg></svg>",
      "isInRegistry": true,
      "name": "Bouilligue",
      "slug": "konnector-bouilligue",
      "type": "konnector",
      "uninstallable": true,
      "versions": Object {
        "beta": Array [
          "0.1.0",
        ],
        "dev": Array [
          "0.1.0",
        ],
        "stable": Array [
          "0.1.0",
        ],
      },
    },
  ],
  "transport": Array [
    Object {
      "categories": Array [
        "transport",
      ],
      "developer": Object {
        "name": "Cozy",
      },
      "icon": "<svg></svg>",
      "installed": true,
      "isInRegistry": true,
      "name": "Trinlane",
      "permissions": Object {
        "mock": Object {
          "type": "io.mock.doctype",
        },
      },
      "slug": "konnector-trinlane",
      "tags": Array [
        "transport",
        "files",
        "bills",
      ],
      "type": "konnector",
      "uninstallable": true,
      "version": "0.1.0",
      "versions": Object {
        "beta": Array [
          "0.1.0",
        ],
        "dev": Array [
          "0.1.0",
        ],
        "stable": Array [
          "0.1.0",
        ],
      },
    },
  ],
}
`;

exports[`sorter should return a categories list alphabetically except for all and others 1`] = `
Array [
  "all",
  "isp",
  "telecom",
  "cozy",
  "others",
]
`;

exports[`sorter should return a categories list alphabetically except for all and others 2`] = `
Array [
  "all",
  "telecom",
  "cozy",
  "others",
]
`;
